File: code_snippets/java/src/main/java/usecases/microservices/Idempotency.java
package usecases.microservices;

import dev.restate.client.Client;

public class Idempotency {

  public static String RESTATE_URL = "http://localhost:8080";

  public void createSubscription() {
    String requestId = "123";
    Utils.SubscriptionRequest subscriptionRequest =
        new Utils.SubscriptionRequest("123", "123", new String[] {"123"});
    // <start_here>
    // <mark_1>
    Client restateClient = Client.connect(RESTATE_URL);
    SubscriptionServiceClient.fromClient(restateClient)
        .send()
        .add(
            subscriptionRequest,
            // <mark_2>
            opt -> opt.idempotencyKey(requestId));
    // </mark_2>
    // </mark_1>
    // <end_here>
  }
}

File: code_snippets/kotlin/src/main/kotlin/usecases/microservices/Idempotency.kt
package usecases.microservices

import dev.restate.sdk.client.CallRequestOptions
import dev.restate.sdk.client.Client
import usecases.microservices.service.SubscriptionServiceClient

class Config {
  companion object {
    const val RESTATE_URL = "http://localhost:8080"
  }
}

class Idempotency {
  suspend fun createSubscription() {
     val requestId = "123"
      val subscriptionRequest = SubscriptionRequest("123", "123", listOf("123"))
      // <start_here>
      val restateClient = Client.connect(Config.RESTATE_URL)
    // <mark_1>
      SubscriptionServiceClient.fromClient(restateClient)
        .send()
        .add(
            subscriptionRequest,
            // <mark_2>
            CallRequestOptions.DEFAULT.withIdempotency(requestId)
            // </mark_2>
        )
    // </mark_1>
  // <end_here>
  }
}
